### PR TITLE
New metering labels

### PR DESCRIPTION
**Pull request by eguzki**
https://issues.redhat.com/browse/THREESCALE-7750

Spec doc: https://docs.google.com/document/d/1bd9VnZ9jKwDBOOKAJltAtvhz3COUYYmRe3-BGE6fDGE/edit#bookmark=id.27cia5h5kj9j

example of one pod
```json
❯ k get pods apicast-production-1-stm67 -o jsonpath="{.metadata.labels}" | jq '.'
{
  "app": "3scale-api-management",
  "deployment": "apicast-production-1",
  "deploymentConfig": "apicast-production",
  "deploymentconfig": "apicast-production",
  "rht.comp": "3scale",
  "rht.comp_ver": "master",
  "rht.prod_name": "Red_Hat_Integration",
  "rht.prod_ver": "master", 
  "rht.subcomp": "apicast-production",
  "rht.subcomp_t": "application",
  "threescale_component": "apicast",
  "threescale_component_element": "production"
}
```

### CODE REVIEW

**Comment by eguzki** [comment]
ready for review @miguelsorianod

File: pkg/3scale/amp/operator/upgrade.go [hunk 1]
```diff
@@ -758,6 +767,103 @@ func (u *UpgradeApiManager) findDeploymentTriggerOnImageChange(triggerPolicies [
...
+
+	deploymentConfigs := []*appsv1.DeploymentConfig{
+		apicast.StagingDeploymentConfig(),
+		apicast.ProductionDeploymentConfig(),
+		backend.ListenerDeploymentConfig(),
+		backend.WorkerDeploymentConfig(),
+		backend.CronDeploymentConfig(),
+		zync.DeploymentConfig(),
+		zync.QueDeploymentConfig(),
+		zync.DatabaseDeploymentConfig(),
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/upgrade.go:800]: This should depend on whether zync-database is set as an external database or not.

Comment (eguzki) [reply pkg/3scale/amp/operator/upgrade.go:800]: good catch. I copy&pasted the upgrade procedure when zync database could not be externalized.

Comment (eguzki) [reply pkg/3scale/amp/operator/upgrade.go]: fixed

File: pkg/3scale/amp/operator/upgrade.go [hunk 2]
```diff
@@ -53,6 +54,14 @@ func (u *UpgradeApiManager) Upgrade() (reconcile.Result, error) {
 		return res, nil
 	}
 
+	res, err = u.upgradePodTemplateLabels()
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/upgrade.go:57]: An obsevation about this: This means that from now on, for every new operator release where any of this label changes all the pods in 3scale will be redeployed.

Comment (eguzki) [reply pkg/3scale/amp/operator/upgrade.go:57]: That is correct

File: pkg/3scale/amp/operator/apicast_options_provider.go [hunk 3]
```diff
@@ -34,11 +34,6 @@ func NewApicastOptionsProvider(apimanager *appsv1alpha1.APIManager, client clien
 }
 
 func (a *ApicastOptionsProvider) GetApicastOptions() (*component.ApicastOptions, error) {
-	imageOpts, err := NewAmpImagesOptionsProvider(a.apimanager).GetAmpImagesOptions()
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/apicast_options_provider.go:37]: Do you think we might actually need this for https://issues.redhat.com/browse/THREESCALE-7481 ?

File: pkg/3scale/amp/operator/redis_options_provider_test.go [hunk 4]
```diff
@@ -223,7 +219,7 @@ func TestGetRedisOptionsProvider(t *testing.T) {
 			func() *component.RedisOptions {
 				opts := defaultRedisOptions()
 				opts.SystemImage = systemRedisImageURL
-				opts.SystemRedisPodTemplateLabels["com.redhat.component-version"] = "systemCustomVersion"
+				//opts.SystemRedisPodTemplateLabels["com.redhat.component-version"] = "systemCustomVersion"
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/redis_options_provider_test.go:222]: This commented line and others can be removed

File: pkg/3scale/amp/operator/upgrade.go [hunk 5]
```diff
@@ -758,6 +767,107 @@ func (u *UpgradeApiManager) findDeploymentTriggerOnImageChange(triggerPolicies [
...
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		updated = updated || updatedTmp
+	}
+
+	return reconcile.Result{Requeue: updated}, nil
+}
+
+func (u *UpgradeApiManager) ensurePodTemplateLabels(desired *appsv1.DeploymentConfig) (bool, error) {
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/upgrade.go:847]: With this changes here we will ensure new labels are added, but the old labels will still be there. Should we clean them? too?

File: pkg/helper/version_parser.go [hunk 6]
```diff
@@ -1,62 +0,0 @@
-package helper
```
Comment (miguelsorianod) [inline pkg/helper/version_parser.go:1]: Same comment as https://github.com/3scale/3scale-operator/pull/679/files#r752285804

File: pkg/helper/labels.go [hunk 7]
```diff
@@ -13,18 +11,15 @@ const (
...
-	}
-
-	if len(validation.IsValidLabelValue(componentVersion)) > 0 {
-		labels["com.redhat.component-version"] = ""
+func MeteringLabels(componentName string, componentType ComponentType) map[string]string {
+	return map[string]string{
+		"com.company":   "Red_Hat",
+		"rht.prod_name": "Red_Hat_Integration",
+		// How to make this version match RHI version? Env var and data from env var?
+		"rht.prod_ver":  "2021.Q4",
```
Comment (eguzki) [inline pkg/helper/labels.go:19]: @miguelsorianod  this `2021.Q4` is specific for the 2.12 release. 

Maybe should I fill with `master` and change in the `3scale-2.12-` branch?

File: pkg/3scale/amp/operator/redis_options_provider_test.go (hunk 4 above)
Comment (eguzki) [reply pkg/3scale/amp/operator/redis_options_provider_test.go]: fixed

File: pkg/3scale/amp/operator/apicast_options_provider.go (hunk 3 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/operator/apicast_options_provider.go:37]: :+1:

File: pkg/helper/labels.go (hunk 7 above)
Comment (miguelsorianod) [reply pkg/helper/labels.go:19]: Seems reasonable to me. We do the same with other labels

File: pkg/3scale/amp/operator/upgrade.go (hunk 5 above)
Comment (eguzki) [reply pkg/3scale/amp/operator/upgrade.go:847]: Added procedure to remove old labels

File: pkg/helper/labels.go (hunk 7 above)
Comment (eguzki) [reply pkg/helper/labels.go]: changed to master

File: pkg/3scale/amp/operator/upgrade.go (hunk 5 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/operator/upgrade.go:847]: This method will need to be reused in future 3scale versions, without the part that removes the old metering labels. If you want you can add a comment to remember to delete that part

File: pkg/3scale/amp/operator/upgrade.go [hunk 8]
```diff
@@ -758,6 +767,119 @@ func (u *UpgradeApiManager) findDeploymentTriggerOnImageChange(triggerPolicies [
...
+	u.Logger().V(1).Info(fmt.Sprintf("ensurePodTemplateLabels object %s", common.ObjectInfo(desired)))
+	existing := &appsv1.DeploymentConfig{}
+	err := u.Client().Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: u.apiManager.Namespace}, existing)
+	if err != nil {
+		return false, err
+	}
+
+	updated := false
+
+	diff := cmp.Diff(existing.Spec.Template.Labels, desired.Spec.Template.Labels)
```
Comment (miguelsorianod) [inline pkg/3scale/amp/operator/upgrade.go:857]: The diff is not 100% accurate now. Maybe the old metering labels and diff calculation should be done now before the merge of the map?

Comment (miguelsorianod) [reply pkg/3scale/amp/operator/upgrade.go:857]: Rethinking about this, I think my previous comment is not correct either. Existing still contains those labels

**Review by miguelsorianod — APPROVED** [review]
I left a minor comment. Looks good to me

File: pkg/3scale/amp/operator/upgrade.go (hunk 5 above)
Comment (eguzki) [reply pkg/3scale/amp/operator/upgrade.go:849]: done

**Pull request merged**